src/graphics.c

```diff
@@ -558,12 +558,46 @@ GdipSaveGraphics (GpGraphics *graphics, unsigned int *state)
 	return Ok;
 }
 
+static GpStatus
+apply_world_to_bounds (GpGraphics *graphics)
+{
+	GpStatus status;
+	GpPointF pts[2];
+
+	pts[0].X = graphics->bounds.X;
+	pts[0].Y = graphics->bounds.Y;
+	pts[1].X = graphics->bounds.X + graphics->bounds.Width;
+	pts[1].Y = graphics->bounds.Y + graphics->bounds.Height;
+	status = GdipTransformMatrixPoints (graphics->clip_matrix, (GpPointF*)&pts, 2);
+	if (status != Ok)
+		return status;
+
+	if (pts[0].X > pts[1].X) {
+		graphics->bounds.X = pts[1].X;
+		graphics->bounds.Width = iround (pts[0].X - pts[1].X);
+	} else {
+		graphics->bounds.X = pts[0].X;
+		graphics->bounds.Width = iround (pts[1].X - pts[0].X);
+	}
+	if (pts[0].Y > pts[1].Y) {
+		graphics->bounds.Y = pts[1].Y;
+		graphics->bounds.Height = iround (pts[0].Y - pts[1].Y);
+	} else {
+		graphics->bounds.Y = pts[0].Y;
+		graphics->bounds.Height = iround (pts[1].Y - pts[0].Y);
+	}
+	return Ok;
+}
+
 GpStatus WINGDIPAPI
 GdipResetWorldTransform (GpGraphics *graphics)
 {
 	if (!graphics)
 		return InvalidParameter;
 
+	GdipInvertMatrix (graphics->clip_matrix);
+	apply_world_to_bounds (graphics);
```

Comment:
The problem here was that each time the clip matrix is updated we use the transform to update the bounds. However, we never did this when calling GdipResetWorldTransform, so the old previously transformed bounds remained.

We need to invert the matrix as we're essentially undoing the previous transformation